packages/cli/src/modules/mcp/tools/execute-workflow.tool.ts

```diff
@@ -0,0 +1,275 @@
+				inputs,
+			);
+
+			telemetryPayload.results = {
+				success: output.success,
+				data: {
+					executionId: output.executionId,
+				},
+			};
+			telemetry.track(USER_CALLED_MCP_TOOL_EVENT, telemetryPayload);
+
+			return {
+				content: [{ type: 'text', text: JSON.stringify(output) }],
+				structuredContent: output,
+			};
+		} catch (error) {
+			const payload: ExecuteWorkflowOutput = {
+				success: false,
+				executionId: null,
+				error,
+			};
+
+			telemetryPayload.results = {
+				success: false,
+				error: error instanceof Error ? error.message : String(error),
+			};
+			telemetry.track(USER_CALLED_MCP_TOOL_EVENT, telemetryPayload);
+
+			return {
+				content: [{ type: 'text', text: JSON.stringify(payload) }],
+				structuredContent: payload,
+			};
+		}
+	},
+});
+
+/**
+ * Executes a workflow for the given user with provided inputs.
+ * In order to "synchronously" execute the workflow,
+ * it is mapping mcp tool inputs to trigger node pin data and starting execution from there.
+ * LIMITATION: Does not properly support workflows with multiple triggers.
+ */
+export const executeWorkflow = async (
+	user: User,
+	workflowFinderService: WorkflowFinderService,
+	activeExecutions: ActiveExecutions,
+	workflowRunner: WorkflowRunner,
+	workflowId: string,
+	inputs?: z.infer<typeof inputSchema>['inputs'],
+): Promise<ExecuteWorkflowOutput> => {
+	const workflow = await workflowFinderService.findWorkflowForUser(workflowId, user, [
+		'workflow:read',
+	]);
+
+	if (!workflow || workflow.isArchived || !workflow.settings?.availableInMCP) {
+		throw new UserError('Workflow not found');
+	}
+
+	const canExecuteWorkflow = isWorkflowEligibleForMCPAccess(workflow);
+
+	if (!canExecuteWorkflow) {
+		throw new UserError(
+			`Only workflows with the following trigger nodes can be executed: ${Object.values(SUPPORTED_MCP_TRIGGERS).join(', ')}.`,
+		);
+	}
+
+	const triggerNode = findMcpWorkflowStart(workflow.nodes);
+
+	if (!triggerNode) {
+		throw new UserError('No supported trigger node found to start the workflow execution.');
+	}
+
+	const runData: IWorkflowExecutionDataProcess = {
+		executionMode: getExecutionModeForTrigger(triggerNode),
+		workflowData: workflow,
+		userId: user.id,
+	};
+
+	// Set the trigger node as the start node and pin data for it
+	// This will enable us to run the workflow from the trigger node with the provided inputs without waiting for an actual trigger event
+	runData.startNodes = [{ name: triggerNode.name, sourceData: null }];
+	runData.pinData = getPinDataForTrigger(triggerNode, inputs ?? {});
+
+	const executionId = await workflowRunner.run(runData);
+
+	if (!executionId) {
+		return {
+			success: false,
+			executionId: null,
+		};
+	}
+	const data = await activeExecutions.getPostExecutePromise(executionId);
+
+	if (data === undefined) {
+		throw new UnexpectedError('Workflow did not return any data');
+	}
+
+	return {
+		success: data.status !== 'error' && !data.data.resultData?.error,
```

Comment:
Why isn't checking the status enough by itself?

Reply:
I saw cases where status was `running` at this point but there was an error in `resultData`. Didn't have a chance to inspect further tho